src/main/java/com/iota/iri/service/ledger/impl/LedgerServiceImpl.java

```diff
@@ -175,6 +175,18 @@ public boolean isBalanceDiffConsistent(Set<Hash> approvedHashes, Map<Hash, Long>
 
                 boolean isEmptyTrunk = trunkTransactionViewModel.getType() == TransactionViewModel.PREFILLED_SLOT;
                 boolean isEmptyBranch = branchTransactionViewModel.getType() == TransactionViewModel.PREFILLED_SLOT;
+
+                //Don't confirm non-solid txs.
+                Hash genesisHash = Hash.NULL_HASH;
+                if(isEmptyTrunk  && !allowGenesisReference){
+                    return null;
+                 }
+                //proceed only if empty trunk or branch is genesis
+                if((isEmptyTrunk && !trunkTransactionViewModel.getHash().equals(genesisHash))||
+                (isEmptyBranch && !branchTransactionViewModel.getHash().equals(genesisHash))){
+                    return null;
+                }
+
```

Comment:
1. In the event that trunk or branch have a non `9s` hash then this is wrong
2. I believe (I think I missed it in the last review so my bad) that the `9s` genesis hash is in the `solidEntryPointsSet`

So I think the flag is unnecessary (because we should always allow genesis reference)
I think we should always return null if it is `PREFILLED_SLOT`

Reply:
Since we return `null` if `branch` or `trunk` is `PREFILLED_SLOT`. 
Is this the only definition of a `leaf`?
```
boolean isLeafTrunk = visitedTransactions.contains(trunkTransactionViewModel.getHash()) || isApprovedTrunk;
```

Reply:
Yes, when `visitedTransactions` contain all SEPs and genesis